### PR TITLE
[Filter/Caffe2] initial commit for caffe2

### PR DESCRIPTION
**Pull request by helloahn**
it will be updated after packaging pytorch properly and example & test also

Signed-off-by: Hyoung Joo Ahn <hello.ahn@samsung.com>

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [*]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

File: ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.cc [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+int
+Caffe2Core::initInputTensor ()
+{
+  int i = 0;
+  inputTensorMap.clear ();
+  for (; i < inputTensorMeta.num_tensors; i++) {
+    Tensor *inputTensor = workSpace.CreateBlob (inputTensorMeta.info[i].name)
+      ->GetMutable<Tensor> ();
+
+  switch (inputTensorMeta.info[i].type){
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.cc]: add space in the loop

File: ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.h [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+#ifndef TENSOR_FILTER_CAFFE2_CORE_H
+#define TENSOR_FILTER_CAFFE2_CORE_H
+
+#include "nnstreamer_plugin_api_filter.h"
+
+#ifdef __cplusplus
+#include <iostream>
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wredundant-decls"
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.h]: Question: ignored "-Wredundant-decls",  is it necessary to avid build warning in caffe2?

File: ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.cc [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+#endif
+  return 0;
+}
+
+/**
+ * @brief extract and store the information of input tensors
+ * @return 0 if OK. non-zero if error.
+ */
+int
+Caffe2Core::setInputTensorProp (const GstTensorFilterProperties * prop)
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.cc]: IMO setInputTensorProp() and setOutputTensorProp() these functions are unnecessary. 
plz consider to copy the meta data in init function.

File: ext/nnstreamer/tensor_filter/tensor_filter_caffe2.c [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+ */
+static int
+caffe2_loadModelFile (const GstTensorFilterProperties * prop,
+    void **private_data)
+{
+  caffe2_data *cf2;
+  if (*private_data != NULL) {
+    /** @todo : Check the integrity of filter->data and filter->model_file, nnfw */
+    cf2 = *private_data;
+    if (strcmp (prop->model_file,
```
Comment (kparichay) [inline ext/nnstreamer/tensor_filter/tensor_filter_caffe2.c]: use `g_strcmp0()`?

File: ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.cc [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+  }
+
+  workSpace.RunNet (predictNet.name ());
+
+  for (i = 0; i < outputTensorMeta.num_tensors; i++) {
+    const auto& out = workSpace.GetBlob (outputTensorMeta.info[i].name)
+      ->Get<Tensor> ();
+    switch (outputTensorMeta.info[i].type){
+      case _NNS_INT32:
+        output[i].data = out.data<int32_t>();
```
Comment (kparichay) [inline ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.cc]: If my understanding is correct, as `allocate_in_invoke` is set to `FALSE`, `GstTensorMemory * output` already has memory allocated, overwriting `output[i].data` will result in loss of that memory.
I think `allocate_in_invoke` should be set to `TRUE`

File: ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.cc [hunk 6]
```diff
@@ -0,0 +1,465 @@
...
+do {\
+  ReinitializeTensor (\
+      inputTensor,\
+      {\
+        inputTensorMeta.info[i].dimension[0],\
+        inputTensorMeta.info[i].dimension[1],\
+        inputTensorMeta.info[i].dimension[2],\
+        inputTensorMeta.info[i].dimension[3]\
+      },\
+      at::dtype<type> ().device (CPU)\
```
Comment (kparichay) [inline ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.cc]: can we infer the device from the model file to be loaded rather than using fixed `CPU`?
Also, what is the default device (if we donot specify any device)?

File: ext/nnstreamer/tensor_filter/tensor_filter_caffe2.c (hunk 4 above)
Comment (myungjoo) [reply ext/nnstreamer/tensor_filter/tensor_filter_caffe2.c]: yup. strcmp might be not that portable especially when a variable may be NULL.

File: ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.cc (hunk 1 above)
Comment (helloahn) [reply ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.cc]: it's added

File: ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.h (hunk 2 above)
Comment (helloahn) [reply ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.h]: it's removed

File: ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.cc (hunk 3 above)
Comment (helloahn) [reply ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.cc]: it's removed

File: ext/nnstreamer/tensor_filter/tensor_filter_caffe2.c (hunk 4 above)
Comment (helloahn) [reply ext/nnstreamer/tensor_filter/tensor_filter_caffe2.c]: it's applied

File: ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.cc (hunk 5 above)
Comment (helloahn) [reply ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.cc]: it's changed.

File: ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.cc (hunk 6 above)
Comment (helloahn) [reply ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.cc]: currently, the default device for every nnfw of nnstreamer is CPU,  because there is no package for gpu, include tf, tf-lite. When the gpu packages are updated, we will make another option for it or respective fw, like tf-gpu, caffe2-gpu.

Comment (kparichay) [reply ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.cc]: ok.

File: ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.cc [hunk 7]
```diff
@@ -0,0 +1,439 @@
...
+  );\
+} while (0);
```
Comment (kparichay) [inline ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.cc]: I dont understand - how does the code flow come out of this loop?

File: ext/nnstreamer/tensor_filter/tensor_filter_caffe2.c [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+    if (g_strcmp0 (prop->model_file,
+            caffe2_core_getModelPath (cf2->caffe2_private_data)) != 0 ||
+        g_strcmp0 (prop->model_file_sub,
+            caffe2_core_getInitModelPath (cf2->caffe2_private_data)) != 0) {
+      caffe2_close (prop, private_data);
+    } else {
+      return 1;
+    }
+  }
+  cf2 = g_new0 (caffe2_data, 1); /** initialize cf2 Fill Zero! */
```
Comment (wooksong) [inline ext/nnstreamer/tensor_filter/tensor_filter_caffe2.c]: Is cf2 freed after returning error?

Comment (helloahn) [reply ext/nnstreamer/tensor_filter/tensor_filter_caffe2.c]: `return 1` doesn't mean an error. It means maintaining current state rather than closing&opening new caffe2 with the new models.

Comment (wooksong) [reply ext/nnstreamer/tensor_filter/tensor_filter_caffe2.c]: @helloahn Do you mean that cf2 would be freed after all in the cases of returning -2 and -1?

Comment (helloahn) [reply ext/nnstreamer/tensor_filter/tensor_filter_caffe2.c]: @wooksong 
I did some tests for that case, and you're right. It's not freed if there is any problem. Specifically, `.close()` is not called.
Furthermore, our nnfws(like tensorflow, tf-lite, pytorch..) have the same problem. I'll fix it at once at another pr.

Comment (wooksong) [reply ext/nnstreamer/tensor_filter/tensor_filter_caffe2.c]: @helloahn 
> Furthermore, our nnfws(like tensorflow, tf-lite, pytorch..) have the same problem.
Right. It is need to be fixed later.

IMHO, in this case, simply add g_free just before ```return -1``` and ```return -2``` would be fixed. Anyway, not a critical issue at this point.
